File: buildercasa.java
import java.util.ArrayList;
import java.util.List;

class Casa {
    private List<String> partes;
    private double costoTotal;

    public Casa() {
        this.partes = new ArrayList<>();
        this.costoTotal = 0.0;
    }

    public void agregar(String parte, double costo) {
        this.partes.add(parte + " - $" + String.format("%.2f", costo));
        this.costoTotal += costo;
    }

    @Override
    public String toString() {
        return "Casa:\n" + String.join("\n", partes) + "\nCosto Total: $" + String.format("%.2f", costoTotal);
    }
}

abstract class ConstructorCasa {
    protected Casa casa;

    public ConstructorCasa() {
        this.casa = new Casa();
    }

    public abstract void construirParedes();
    public abstract void construirTecho();
    public abstract void construirPuertas();
    public abstract void construirVentanas();

    public void reiniciar() {
        this.casa = new Casa();
    }

    public Casa obtenerCasa() {
        return this.casa;
    }
}

class ConstructorCasaMadera extends ConstructorCasa {
    public void construirParedes() {
        casa.agregar("Paredes de madera", 5000);
    }

    public void construirTecho() {
        casa.agregar("Techo de madera", 3000);
    }

    public void construirPuertas() {
        casa.agregar("Puertas de madera", 1200);
    }

    public void construirVentanas() {
        casa.agregar("Ventanas de vidrio", 2000);
    }
}

class ConstructorCasaLadrillo extends ConstructorCasa {
    public void construirParedes() {
        casa.agregar("Paredes de ladrillo", 8000);
    }

    public void construirTecho() {
        casa.agregar("Techo de tejas", 5000);
    }

    public void construirPuertas() {
        casa.agregar("Puertas de metal", 2000);
    }

    public void construirVentanas() {
        casa.agregar("Ventanas de vidrio reforzado", 2500);
    }
}

class DirectorCasa {
    private ConstructorCasa constructor;

    public DirectorCasa(ConstructorCasa constructor) {
        this.constructor = constructor;
    }

    public void construirCasaCompleta() {
        constructor.reiniciar();
        constructor.construirParedes();
        constructor.construirTecho();
        constructor.construirPuertas();
        constructor.construirVentanas();
    }

    public void construirCasaBasica() {
        constructor.reiniciar();
        constructor.construirParedes();
        constructor.construirTecho();
    }
}

public class buildercasa {
    public static void main(String[] args) {
        
        ConstructorCasa constructorMadera = new ConstructorCasaMadera();
        DirectorCasa directorMadera = new DirectorCasa(constructorMadera);
        directorMadera.construirCasaCompleta();
        System.out.println(constructorMadera.obtenerCasa());

        
        ConstructorCasa constructorLadrillo = new ConstructorCasaLadrillo();
        DirectorCasa directorLadrillo = new DirectorCasa(constructorLadrillo);
        directorLadrillo.construirCasaCompleta();
        System.out.println(constructorLadrillo.obtenerCasa());
    }
}

File: buildercasa.py
class Casa:
    def __init__(self):
        self.partes = []
    
    def agregar(self, parte):
        self.partes.append(parte)
    
    def __str__(self):
        return "Casa:\n" + "\n".join(self.partes)
    def get_total_cost(self):
        return sum(item.get_cost() for item in self.items)


class ConstructorCasa:
    def __init__(self):
        self.casa = Casa()
    
    def construir_paredes(self):
        pass
    
    def construir_techo(self):
        pass
    
    def construir_puertas(self):
        pass
    
    def construir_ventanas(self):
        pass
    
    def reiniciar(self):
        self.casa = Casa()
    
    def obtener_casa(self):
        return self.casa


class ConstructorCasaMadera(ConstructorCasa):
    def construir_paredes(self):
        self.casa.agregar("Paredes de madera",0.85)
    
    def construir_techo(self):
        self.casa.agregar("Techo de madera",15)
    
    def construir_puertas(self):
        self.casa.agregar("Puertas de madera",31)
    
    def construir_ventanas(self):
        self.casa.agregar("Ventanas de vidrio",30)


class ConstructorCasaLadrillo(ConstructorCasa):
    def construir_paredes(self):
        self.casa.agregar("Paredes de ladrillo",0.85)
    
    def construir_techo(self):
        self.casa.agregar("Techo de tejas",15)
    
    def construir_puertas(self):
        self.casa.agregar("Puertas de metal",34)
    
    def construir_ventanas(self):
        self.casa.agregar("Ventanas de vidrio reforzado",30)


class DirectorCasa:
    def __init__(self, constructor):
        self.constructor = constructor
    
    def construir_casa_completa(self):
        self.constructor.reiniciar()
        self.constructor.construir_paredes()
        self.constructor.construir_techo()
        self.constructor.construir_puertas()
        self.constructor.construir_ventanas()
    
    def construir_casa_basica(self):
        self.constructor.reiniciar()
        self.constructor.construir_paredes()
        self.constructor.construir_techo()


# Construcción y prueba de una casa de madera
constructor_madera = ConstructorCasaMadera()
director_madera = DirectorCasa(constructor_madera)
director_madera.construir_casa_completa()
print(constructor_madera.obtener_casa())

# Construcción y prueba de una casa de ladrillo
constructor_ladrillo = ConstructorCasaLadrillo()
director_ladrillo = DirectorCasa(constructor_ladrillo)
director_ladrillo.construir_casa_completa()
print(constructor_ladrillo.obtener_casa())
